### PR TITLE
bpo-27129: Use instruction offsets, not byte offsets, in bytecode and internally.

### PR DESCRIPTION
**Pull request by markshannon**
* Changes all offsets in jumps and branches to instruction from byte offsets. This doubles the range of instruction that can be targeted by an instruction.
* Change the meaning of the C field `f_lasti` to match, which saves a bit of shifting in the interpreter.

https://bugs.python.org/issue27129

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -4069,7 +4064,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
         }
 
         case TARGET(SETUP_FINALLY): {
-            PyFrame_BlockSetup(f, SETUP_FINALLY, INSTR_OFFSET() + oparg,
+            PyFrame_BlockSetup(f, SETUP_FINALLY, (next_instr - first_instr) + oparg,
```
Comment (erlend-aasland) [inline Python/ceval.c]: Why this change? `INSTR_OFFSET()` expands to `((int)(next_instr - first_instr))`

Ditto for lines 4101 and 4129.

File: Objects/frameobject.c [hunk 2]
```diff
@@ -56,6 +55,15 @@ frame_getlineno(PyFrameObject *f, void *closure)
     return PyLong_FromLong(PyFrame_GetLineNumber(f));
 }
 
+static PyObject *
+frame_getlasti(PyFrameObject *f, void *closure)
+{
+    if (f->f_lasti < 0) {
+        return PyLong_FromLong(-1);
```
Comment (erlend-aasland) [inline Objects/frameobject.c:62]: I'm wondering if adding `_PyLong_GetMinusOne` to `Include/internal/pycore_long.h` and using that here (`Py_NewRef(_PyLong_GetMinusOne)`) would make a difference. I don't know how hot this branch is, so for all I know, it wouldn't make a difference.

File: Python/ceval.c (hunk 1 above)
Comment (markshannon) [reply Python/ceval.c:4067]: `INSTR_OFFSET()` is the offset in bytes, but we are now using instruction offsets internally.

File: Objects/frameobject.c (hunk 2 above)
Comment (markshannon) [reply Objects/frameobject.c:62]: This is not hot code at all.

File: Python/ceval.c (hunk 1 above)
Comment (markshannon) [reply Python/ceval.c]: I suspect that I changed `INSTR_OFFSET()` after changing this code.
I'll change it back now that `INSTR_OFFSET()` has been updated.

File: Python/ceval.c [hunk 3]
```diff
@@ -6475,7 +6470,7 @@ dtrace_function_entry(PyFrameObject *f)
     PyCodeObject *code = f->f_code;
     filename = PyUnicode_AsUTF8(code->co_filename);
     funcname = PyUnicode_AsUTF8(code->co_name);
-    lineno = PyCode_Addr2Line(code, f->f_lasti);
+    lineno = PyCode_Addr2Line(code, f->f_lasti*2);
```
Comment (erlend-aasland) [inline Python/ceval.c]:
```suggestion
    lineno = PyFrame_GetLineNumber(f);
```

File: Python/ceval.c [hunk 4]
```diff
@@ -6490,7 +6485,7 @@ dtrace_function_return(PyFrameObject *f)
...
```
Comment (erlend-aasland) [inline Python/ceval.c]:
```suggestion
    lineno = PyFrame_GetLineNumber(f);
```

**Review by erlend-aasland — COMMENTED** [review]
One more thing: I'd swap the two `PyCode_Addr2Line`'s in `Python/ceval.c` with `PyFrame_GetLineNumber` for improved readability. (Unless the extra function call impacts performance.)

File: Python/traceback.c [hunk 5]
```diff
@@ -763,8 +763,7 @@ dump_frame(int fd, PyFrameObject *frame)
         PUTS(fd, "???");
     }
 
-    /* PyFrame_GetLineNumber() was introduced in Python 2.7.0 and 3.2.0 */
```
Comment (vstinner) [inline Python/traceback.c:766]: What a strange comment :-D I read it as "oh, Python 2.7 is too young, let's wait a little bit longer before we can use this shiny new function" !?

File: Objects/codeobject.c [hunk 6]
```diff
@@ -1246,7 +1246,7 @@ PyTypeObject PyCode_Type = {
 int
 PyCode_Addr2Line(PyCodeObject *co, int addrq)
```
Comment (vstinner) [inline Objects/codeobject.c:1247]: I would prefer to treat the second argument as an instruction offset, rather than a byte offset. Rather than changing PyCode_Addr2Line() calls. It sounds unfortunate that the API allows to ask the address in the middle of a word (2 bytes).

It sounds common to pass f_lasti to PyCode_Addr2Line(): PyCode_Addr2Line(code, f_lasti) would still work, no?

**Review by vstinner — COMMENTED** [review]
Oh... I didn't notice that when the wordcode work has been done. It's unfortunate that we didn't switch to instruction offset. I vaguely recall discussion about sizeof(_Py_CODEUNIT).

I support this change, here are some comments. My main worry is about PyCode_Addr2Line().

cc @serhiy-storchaka

File: Objects/codeobject.c (hunk 6 above)
Comment (markshannon) [reply Objects/codeobject.c:1247]: Depends on how you think the API is defined.
If you want the the line number from an offset derived from `code.co_linetable`, then the API should be unchanged.

Comment (markshannon) [reply Objects/codeobject.c:1247]: `PyCode_Addr2Line(code, offset)` must take a *byte* offset. 
To get`f_lasti` from the frame, the only API to do so is something like `PyObject_GetAttrString(frame, "f_lasti")`, which would return a byte offset. 

Also see:
https://www.python.org/dev/peps/pep-0626/#c-api
https://www.python.org/dev/peps/pep-0626/#out-of-process-debuggers-and-profilers

File: Python/ceval.c [hunk 7]
```diff
@@ -5511,7 +5506,7 @@ call_trace(Py_tracefunc func, PyObject *obj,
     }
     else {
         initialize_trace_info(trace_info, frame);
-        frame->f_lineno = _PyCode_CheckLineNumber(frame->f_lasti, &trace_info->bounds);
+        frame->f_lineno = _PyCode_CheckLineNumber(frame->f_lasti*2, &trace_info->bounds);
```
Comment (erlend-aasland) [inline Python/ceval.c:5509]: `_PyCode_CheckLineNumber` implies that it takes an instruction offset as its first argument, but in reality it's a bytecode offset. I suggest either changing `_PyCode_CheckLineNumber` to take a bytecode offset:

```diff
diff --git a/Objects/codeobject.c b/Objects/codeobject.c
index 69cb31c1ab..8fd65b48fc 100644
--- a/Objects/codeobject.c
+++ b/Objects/codeobject.c
@@ -1274,16 +1274,17 @@ _PyCode_InitAddressRange(PyCodeObject* co, PyCodeAddressRange *bounds)
 }
 
 /* Update *bounds to describe the first and one-past-the-last instructions in
-   the same line as lasti.  Return the number of that line, or -1 if lasti is out of bounds. */
+   the same line as the bytecode index, addrq.  Return the number of that line,
+   or -1 if lasti is out of bounds. */
 int
-_PyCode_CheckLineNumber(int lasti, PyCodeAddressRange *bounds)
+_PyCode_CheckLineNumber(int addrq, PyCodeAddressRange *bounds)
 {
-    while (bounds->ar_end <= lasti) {
+    while (bounds->ar_end <= addrq) {
         if (!PyLineTable_NextAddressRange(bounds)) {
             return -1;
         }
     }
-    while (bounds->ar_start > lasti) {
+    while (bounds->ar_start > addrq) {
         if (!PyLineTable_PreviousAddressRange(bounds)) {
             return -1;
         }

```

... or altering it to treat its first argument as an instruction offset (and update the callers):

```diff
diff --git a/Objects/codeobject.c b/Objects/codeobject.c
index 69cb31c1ab..41ace1c404 100644
--- a/Objects/codeobject.c
+++ b/Objects/codeobject.c
@@ -1278,12 +1278,13 @@ _PyCode_InitAddressRange(PyCodeObject* co, PyCodeAddressRange *bounds)
 int
 _PyCode_CheckLineNumber(int lasti, PyCodeAddressRange *bounds)
 {
-    while (bounds->ar_end <= lasti) {
+    int addrq = lasti * 2;  // Bytecode index.
+    while (bounds->ar_end <= addrq) {
         if (!PyLineTable_NextAddressRange(bounds)) {
             return -1;
         }
     }
-    while (bounds->ar_start > lasti) {
+    while (bounds->ar_start > addrq) {
         if (!PyLineTable_PreviousAddressRange(bounds)) {
             return -1;
         }

```

**Comment by vstinner** [comment]
I wanted to ask to document properly that PyCode_Addr2Line() uses a byte offset, but the function is not documented! Can you please document the function at:
https://docs.python.org/dev/c-api/code.html

Please document f_lasti unit and the change at: https://docs.python.org/dev/library/inspect.html

**Comment by markshannon** [comment]
The comment on `PyCode_Addr2Line()` says 
"If you just need the line number of a frame, use PyFrame_GetLineNumber() instead."
PEP 626 provides a faster way to extract line numbers: https://www.python.org/dev/peps/pep-0626/#out-of-process-debuggers-and-profilers
So I think we should keep `PyCode_Addr2Line()` undocumented, unless you want to document it as "don't use this."

There are no changes to the inspect module.

File: Lib/importlib/_bootstrap_external.py [hunk 8]
```diff
@@ -314,6 +314,7 @@ def _write_atomic(path, data, mode=0o666):
 #     Python 3.10a2 3432 (Function annotation for MAKE_FUNCTION is changed from dict to tuple bpo-42202)
 #     Python 3.10a2 3433 (RERAISE restores f_lasti if oparg != 0)
 #     Python 3.10a6 3434 (PEP 634: Structural Pattern Matching)
+#     Python 3.10a6 3435 Use instruction offsets (as opposed to byte offsets).
```
Comment (gvanrossum) [inline Lib/importlib/_bootstrap_external.py]: a7?

File: Lib/dis.py [hunk 9]
```diff
@@ -338,8 +338,11 @@ def _get_instructions_bytes(code, varnames=None, names=None, constants=None,
                 argval, argrepr = _get_const_info(arg, constants)
             elif op in hasname:
                 argval, argrepr = _get_name_info(arg, names)
+            elif op in hasjabs:
+                argval = arg*2
+                argrepr = "to " + repr(argval)
```
Comment (gvanrossum) [inline Lib/dis.py:343]: Why not go a step further and divide the labels in dis output by 2?

Comment (markshannon) [reply Lib/dis.py:343]: A couple of reasons.
* I wanted to keep changes to a minimum.
* There is no guarantee that instructions will remain the same size, so using byte offsets keeps things more consistent across versions.

File: Lib/importlib/_bootstrap_external.py (hunk 8 above)
Comment (markshannon) [reply Lib/importlib/_bootstrap_external.py]: Depends on how long it takes to get this merged :smile: 
We are in alpha 6 now.

**Comment by markshannon** [comment]
@vstinner I'm fine with documenting `PyCode_Addr2Line()` but it hasn't been changed in this PR, so that is out of scope for this PR. I'll make another PR.

File: Lib/importlib/_bootstrap_external.py (hunk 8 above)
Comment (vstinner) [reply Lib/importlib/_bootstrap_external.py]: No, alpha6 was already release. The next one is alpha7:
* https://www.python.org/downloads/release/python-3100a6/
* https://www.python.org/dev/peps/pep-0619/

File: Lib/dis.py (hunk 9 above)
Comment (gvanrossum) [reply Lib/dis.py:343]: Sure. However now we have a confusing discrepancy between what's displayed and how we're encouraged to think about the bytecode in the C code. I guess there's no perfect answer, so status quo wins.

**Comment by vstinner** [comment]
> @markshannon markshannon closed this 8 hours ago
> @markshannon markshannon reopened this 8 hours ago 

You should be allowed to click on the CI to "Re-run all jobs" to avoid closing/reopening the PR. Sadly, it's not possible to only re-run a single job.
